proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+
+        module Top where
+          module A where
+            x = 5
+          module B where
+            x = True
+
+      In this example, the ambient QEnv for ``Top`` includes mappings
+      ``{ A.x ↦ Top.A.x, x ↦ Top.A.x, B.x ↦ Top.B.x, x ↦ Top.B.x }``.
+      This QEnv contains two mappings for ``x``. By itself, this does not cause
+      a problem: it is only a usage site of ``x`` that would cause an ambiguity
+      error (as usual). However, since ``Top`` does not have an export list,
+      we would normally export all of the ambient QEnv in ``Top``, causing trouble
+      for importing modules.
+
+      We thus exclude from the default export list any qualified name that is mapped to multiple
+      original names. More formally::
+
+        X' = X \ { qual ↦ orig | qual ↦ orig ∈ X
+                               , qual ↦ orig' ∈ X
+                               , orig /= orig' }
+
+      Now, ``X'`` is our final export QEnv.
+
+   #. **Examples**. For instance in
+
+      ::
+
+        module M (module qualified X (f)) where
+
+          module qualified X where
+            f = …
+            g = …
+
+      the module ``M`` exports ``{X.f ↦ (M, X.f)}``. And in
+
+      ::
+
+       module M (module qualified X) where
+
+         module qualified X where
+           f = …
+           g = …
+
+      the module ``M`` exports ``{X.f ↦ (M, X.f), X.g ↦ (M, X.g)}``. And in
+
+      ::
+
+         module X where
+           module Y where
+             f = …
+             g = …
+
+         module M (module X) where
+           import X
+
+      the export environment of ``M`` is ``{f ↦ (X, Y.f), g ↦ (x, Y.g)}``.
+
+#. **Local Modules**
+
+   A. Introduce a new declaration form
+      to declare a new module called a *local module*. Here is the BNF::
+
+        topdecl ::= ... | 'module' [ 'qualified' ] modids [ exports ] 'where' '{' topdecls '}'
+
+      This declaration form is allowed only with ``-XLocalModules``.
+
+      The ``module`` keyword in a local module declaration must not be the first
+      lexeme in a file. (This is to avoid ambiguity with top-level,
+      file-sized modules.)
+
+      Note that this new form is part of ``topdecl``, meaning that it cannot appear
+      in a ``let`` or ``where``. However, because a local module contains ``topdecl``\ s,
+      local modules may be arbitrarily nested.
+
+   #. There are now two kinds of modules: top-level modules and local
+      modules. Local modules lack import statements, but can define all
+      the range of entities definable in a top-level module.
+
+      The term *module* can refer to either a top-level or a local module.
+
+   #. The ambient QEnv within the local module extends the ambient QEnv
+      of the enclosing module with its local definitions. This means that
+      all qualified names in scope in the enclosing module are in scope in the
+      local module.
+
+   #. Definitions in a local module may be mutually recursive with definitions
```

Comment:
No. My point here isn't so much as to allow this new feature, but to describe the whole modules mechanism as existing entirely in the renamer. It's really just about names. So, because any arbitrary Haskell definitions can be mutually recursive, so can definitions in different local modules. Put another way, I think allowing this sort of mutual recursion *simplifies* the extension, not complexifies it.